### PR TITLE
Proof of concept tests

### PR DESCRIPTION
**Pull request by jhrcek**
Here's a proof of concept of some basic integration tests.
Before I continue writing any more test I'd like to know your opinion about this approach.

Apart from comments about the approach I chose, I'd also welcome couple of pointers about what kinds of tests 
(given X I expect Y to happen) you'd like to find most useful.

### CODE REVIEW

File: hiedb.cabal [hunk 1]
```diff
@@ -57,3 +57,28 @@ library
...
+    build-depends:      base >= 4.12 && < 4.15
+                      , directory
+                      , hspec
+                      , process
+                      , temporary
+                      , test-lib
+    default-language:   Haskell2010
+
+-- Private library used in tests
+library test-lib
```
Comment (jhrcek) [inline hiedb.cabal:77]: [internal library](https://cabal.readthedocs.io/en/3.4/cabal-package.html#sublibs) - only tests will see this.
Tests depend on this, so it will be compiled before tests.
Tests will use the .hie dir produced by ghc-options below as the base directory for indexing

File: test-lib/Foo.hs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+module Foo
```
Comment (jhrcek) [inline test-lib/Foo.hs:1]: Just some dummy modules for now. We could expand it to something more complex to test various edge case.

File: hiedb.cabal [hunk 3]
```diff
@@ -57,3 +57,28 @@ library
   hs-source-dirs:      src
   default-language:    Haskell2010
   ghc-options:         -Wall
+
+test-suite hiedb-tests
+    type:               exitcode-stdio-1.0
+    hs-source-dirs:     test
+    main-is:            Main.hs
+    build-tool-depends: hiedb:hiedb
```
Comment (jhrcek) [inline hiedb.cabal:65]: A trick I found on stack overflow - using this the test suite depends on hiedb binary built before tests are executed `cabal test`.

File: test/Main.hs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+          runHieDbCli ["index" , ".hie", "--quiet"]
+            `suceedsWithStdin` ""
+
+      describe "ls" $
+        it "lists the indexed modules" $ do
+          runHieDbCli ["ls"]
+            `succeedsWithStdinSatisfying` (\stdin -> do
+              let inlines = lines stdin
+              length inlines `shouldBe` 2
+              (inlines !! 0) `shouldSatisfy` isSuffixOf ".hie/Foo.hie\tFoo\thiedb-0.1.0.0-inplace-test-lib"
```
Comment (jhrcek) [inline test/Main.hs:28]: Not suer how brittle these kinds of tests will be.
E.g. here I had to do this trick with custom predicates, because the indexed hie file paths are full paths which will differe from system to system.

**Comment by wz1000** [comment]
I don't think hackage is very happy with internal libraries. I think there was something wrong with the stack support too.

I think we should have these files in some folder `test/data`, and call ghc explicitly `ghc -fwrite-ide-info A.hs B.hs ...` before running any tests, instead of having them be part of a cabal project.

File: test/Main.hs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import System.Exit (ExitCode (ExitSuccess), die)
+import System.Process (proc, readCreateProcessWithExitCode)
+import Test.Hspec (Expectation, Spec, afterAll_, describe, hspec, it, shouldBe, shouldSatisfy)
+
+main :: IO ()
+main = hspec spec
+
+spec :: Spec
+spec = do
+  describe "hiedb" $ do
```
Comment (wz1000) [inline test/Main.hs:15]:
```suggestion
  describe "hiedb (command line)" $ do
```

So ideally we'd have `"hiedb (api)"` later on

**Comment by wz1000** [comment]
Generally seems like a good start!

File: .gitignore [hunk 6]
```diff
@@ -2,3 +2,4 @@
 dist-newstyle/
 cabal.project.local*
 *.buildinfo
+.hie
```
Comment (wz1000) [inline .gitignore:5]: maybe we want to add `*.hiedb` too

Comment (jhrcek) [reply .gitignore]: I switched to putting all test output (ghc output, like .hi, .o and .hie files) into `test/tmp` which is removed after tests, so shouldn't need to add anything to .gitignore

File: test/Main.hs (hunk 4 above)
Comment (jhrcek) [reply test/Main.hs]: It will be probably best to switch to some form of golden testing, with expected stdout / stderr for each command run in separate file. I'll look into it next

File: hiedb.cabal (hunk 1 above)
Comment (jhrcek) [reply hiedb.cabal]: switched away from internal library to calling ghc directly. I assume it's ok for the test to use the "ghc" currently present on PATH. or would you suggest some other approach to locating ghc?

**Comment by jhrcek** [comment]
Today I realized I can use the API in the tests directly :D I was so focused on the CLI that I didn't realize at all, that hiedb is a library with public modules that I can use in the tests. Working on API tests now and marking this as draft before I have something more substantial..

File: test/Test/Orphans.hs [hunk 7]
```diff
@@ -0,0 +1,19 @@
+{-# LANGUAGE StandaloneDeriving #-}
+module Test.Orphans where
+
+import HieDb.Types
+import Module (ModuleName, moduleName, moduleNameString, moduleUnitId)
+import Name (Name, nameModule, nameOccName)
+import OccName (OccName, occNameString)
+
+instance Show ModuleName where show = moduleNameString
```
Comment (jhrcek) [inline test/Test/Orphans.hs:9]: I'm a bit unsure about how to approach testing with these types.
Most hspec combinators (like shouldBe, shouldReturn) depends on Show and Eq instance, but most of these GHC types don't have one. Is it ok to derive bunch of orphan instances just for the purpose of testing, or how do GHC devs approach this?

**Comment by jhrcek** [comment]
@wz1000 Please check before I continue writing more API tests. I'm a bit unsure how to handle missing Show/Eq instance for those GHC types.

File: test/Test/Orphans.hs (hunk 7 above)
Comment (wz1000) [reply test/Test/Orphans.hs:9]: Try projecting out parts which have the instances? Like `moduleNameString`.

If this proves to be insufficient, you can define orphans in the test suite.

File: test/Main.hs [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+cleanTestData :: IO ()
+cleanTestData = removeDirectoryRecursive testTmp
+
+compileTestModules :: IO ()
+compileTestModules =
+  callProcess "ghc" $
+    "-fwrite-ide-info" :
+    "-hiedir=" <> testTmp :
+    "-hidir=" <> testTmp :
```
Comment (wz1000) [inline test/Main.hs:145]: You can use `-fnocode` to get rid of `.o` and `.hi` files.

**Comment by jhrcek** [comment]
@wz1000 I consider this review / merge ready.
There's not many API tests, but more than 2/3 of CLI commands are covered by at least 1 end to end tests.
I'll be writing more API tests next, but first I'd like to get this merged as intermediate progress before I implement the removal of unnecessary file field from the tables. I tested this locally with GHC 8.10.2 and 8.8.4.

**Comment by wz1000** [comment]
Very nice. Great work!

**Comment by wz1000** [comment]
> I tested this locally with GHC 8.10.2 and 8.8.4 - but not with 8.6.5 (as it doesn't have -fwrite-ide-info and not sure how the testing would work with that).

I think we can use `hie-compat` and the GHC api to generate the files, but don't worry about this for now.

**Review by wz1000 — APPROVED** [review]

**Comment by wz1000** [comment]
BTW, we can use something like `cabal get $package` to download some non-trivial packages to index.
